### PR TITLE
Unsized rvalues: implement boxed closure impls.

### PR DESCRIPTION
**Pull request by qnighy**
This pull request contains **`boxed_closure_impls`** that provides long-hoped three impls:

```rust
impl<A, F: FnOnce<A> + ?Sized> FnOnce<A> for Box<F> { .. }
impl<A, F: FnMut<A> + ?Sized> FnMut<A> for Box<F> { .. }
impl<A, F: Fn<A> + ?Sized> Fn<A> for Box<F> { .. }
```

This has been blocked by several reasons; see `FnBox` #28796 for details.

Now that #48055 is (partly) implemented, we're ready to introduce the above impls, replacing existing `FnBox` workarounds.

There are two major concerns, however.

## Major concern 1: instability

I think these impls should be introduced as an unstable feature first, mainly because it relies on unsized rvalues. However, `impl` itself can't stand as unstable; all existing unstable features are tied with functions, methods, or types. I tried putting `#[unstable]` on the impls but that didn't work. <del>**I'll mark this PR as [WIP] until it is resolved**</del>. I'll just remove `[WIP]` and let the compiler team decide how to deal with the instability.

## Major concern 2: compatibility with `FnBox`

I'm not really sure, but `FnBox` may be widely used in the nightly world. Although unstable features may regress, it's good if there's a path for gradual migration. It can be done in the following way:

- Make `FnBox` a subtrait of `FnOnce`. This ensures that `dyn FnBox` implements `FnOnce`.
- Make use of specialization to avoid overlap between `FnOnce` impls for `Box<impl FnOnce>` and `Box<dyn FnBox>`.

I believe that this minimizes breakage of crates that use `FnBox`.

## Minor concern: feature name and tracking issue

I currently assign `boxed_closure_impls` as the name and #48055 as the tracking issue. I'll prepare a separate tracking issue when the Major Concern 1 is resolved.

### CODE REVIEW

**Comment by bjorn3** [comment]
@TimNN ^ not very helpful

**Comment by alexreg** [comment]
`FnBox` was only ever a stop-gap solution, and highly unstable. I really wouldn't worry about migration, and would just obsolete it immediately in favour of this, if it works just as well/better in all circumstances.

**Comment by Mark-Simulacrum** [comment]
I've re-assigned this to @nikomatsakis since I think it'll probably want lang/compiler team discussion or review.

**Comment by Aaronepower** [comment]
Triage; @nikomatsakis Hello, have you been able to get back to this PR?

**Comment by Dylan-DPC** [comment]
Pinging from triage @rust-lang/compiler can anyone review this?

**Comment by eddyb** [comment]
This is *not really* a compiler change. Maybe @rust-lang/lang + @rust-lang/libs?

**Comment by cramertj** [comment]
@eddyb I think it is compiler-relevant since one of the core questions in the PR description is how to make use of the new impls unstable:

> However, impl itself can't stand as unstable; all existing unstable features are tied with functions, methods, or types. I tried putting #[unstable] on the impls but that didn't work. I'll mark this PR as [WIP] until it is resolved.

This isn't possible today, and it's hard for me to understand how it could *become* possible: we'd have to somehow gate all coercions, where clauses, transitive uses (e.g. `trait MyTrait {} impl<T> MyTrait for T where T: FnOnce() { ... }`) etc. I'm personally fine exposing this from the lang side as it seems obvious that these impls *should* exist somehow. The only hold-up I would have is if we accidentally make it backwards-incompatible to change the impl of `FnOnce` for `Box<T>` to avoid alloca by passing `self` by-pointer and dropping the allocated memory after the call. I can't think of many reasonable ways that a user could be relying on the alloca + move-to-stack occurring, though, so I think this is fine and I don't see a reason not to land these impls now that we finally *can*.

**Comment by eddyb** [comment]
@cramertj I think that's a lang team discussion topic - that is, whether we should just stabilize now, given that feature-gating is hard (or even impossible).
Maybe we should start an FCP merge on this PR?

**Comment by cramertj** [comment]
@rfcbot fcp merge

This change introduces the following (insta-stable) trait impls:
```rust
impl<A, F: FnOnce<A> + ?Sized> FnOnce<A> for Box<F> { .. }
impl<A, F: FnMut<A> + ?Sized> FnMut<A> for Box<F> { .. }
impl<A, F: Fn<A> + ?Sized> Fn<A> for Box<F> { .. }
```
Defining these impls requires the use of the `unsized_locals` feature and we'd ideally like to gate these impls since they rely on unstable features in order to make it possible to implement them (i.e. it's not just this particular impl that is unstable as with e.g. most of std, but implementing them at all is impossible without the unstable feature or some extra-clever and probably UB hacks).

However, we don't have the machinery (and possibly won't ever have the machinery) to make trait impls unstable. As such, merging this PR makes `Box<FnOnce()>` instantly call-able on stable.

These implementations have been desired for a long time but have been impossible to add until now. I'm not sure of any other reasonable ways to implement them, or any ways that we might want to change the implementations going forward in a user-observable way.

**Comment by Centril** [comment]
Is there a particular reason to stabilize these impls now rather than to wait until `unsized_locals` is stable?

**Comment by alexreg** [comment]
@Centril The surface area is much smaller for these three impls, I guess, and they cover a large portion of the urgent demand for unsized locals at the same time.

**Comment by alexcrichton** [comment]
@rfcbot concern unimplementable-if-unsized-does-not-pan-out

Historically we've avoiding expanding the stable API surface area of the standard library with unstable features. One of the best examples here is specialization where we've exclusively used it so far for performance improvements and we haven't started to use it for actually expanding trait impls yet. The rationale for this is that we can probably recover performance via other means if absolutely necessary if the unstable feature (specialization) doesn't pan out.

Do we have a similar recourse for this? What if the unstable feature doesn't pan out? Are we confident enough that we can provide this implementation for all of time?

**Comment by cramertj** [comment]
@alexcrichton So I'd argue this *is* implementable even if the current unsized-rvalues-using-alloca proposal doesn't shake out-- we can still pass `self` by-pointer into the method and clean up the memory for the `Box` after the fact, or add a vtable shim to `FnOnce` specifically that takes `Box<Self>`. I think we definitely want `self`-by-value to be object-safe for all the benefits it gives, so I'd argue we want something that can provide this functionality regardless of whether the current feature as-implemented shakes out.

Compared with specialization, where exactly which implementations are allowed and when overlaps may occur is still open to soundness-related debate, the likelihood of the API surface needing to change here is far smaller. Basically the only thing that's being promised here is "there is some way to call `FnOnce::call_once` on a `Box<dyn FnOnce()>`". I think there are enough alternatives (pass-by-pointer, a special vtable shim which takes `Box<Self>`, etc.) that we can feel safe providing this API.

**Comment by cramertj** [comment]
@Centril 
> Is there a particular reason to stabilize these impls now rather than to wait until unsized_locals is stable?

`Box<dyn FnOnce(..) -> ...>` working would be a massive improvement to the ergonomics of lots of code which currently relies on [hacky custom traits for every function signature needed](https://play.rust-lang.org/?version=stable&mode=debug&edition=2015&gist=bc42a3acf895a20a480f9ad089dc01f3). As I said above in my response to @alexcrichton, I feel like the API surface being provided here is high value, small, and low-risk enough that it needn't be blocked by the much larger and more complex `unsized_locals` feature.

**Comment by alexcrichton** [comment]
@rfcbot resolve unimplementable-if-unsized-does-not-pan-out

Sounds solid to me, thanks for the explanation!

File: src/doc/unstable-book/src/library-features/boxed-closure-impls.md [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+        // *self is an unsized rvalue
+        <F as FnOnce<A>>::call_once(*self, args)
+    }
+}
+```
+
+However, since `unsized_locals` is a very young feature, we're careful about
+this `FnOnce` impl now.
+
+There's another reason for instability: for compatibility with `fnbox`,
```
Comment (cramertj) [inline src/doc/unstable-book/src/library-features/boxed-closure-impls.md]: There isn't a way for users to observe this on stable, and IMO we should deprecate and remove `FnBox` as soon as possible so we can remove the `default` here.

**Comment by dtolnay** [comment]
Just to confirm -- this is also going to work for `Box<FnOnce() + Send>`, `Box<FnOnce() + Sync>`, `Box<FnOnce() + Send + Sync>` right?

**Comment by Centril** [comment]
@rfcbot concern i-need-to-think-about-this

**Comment by cramertj** [comment]
@Centril 
> @rfcbot concern i-need-to-think-about-this

Have you had a chance to think about this? Also, ping 
 @SimonSapin
  @alexcrichton
  @aturon
  @joshtriplett
  @nikomatsakis 
  @pnkfelix
  @scottmcm

Y'all have unchecked boxes :)

**Comment by Centril** [comment]
@cramertj 

> Have you had a chance to think about this?

@rfcbot resolve i-need-to-think-about-this
@rfcbot concern risky-stabilization-order

Yeah; I'm nervous about landing an implementation relying on a type system change that isn't stable yet; I think generally, we should move more deliberately with type system features as compared to other areas of the language. Once it is stable I agree we'll want to land these impls for sure. If `unsized_locals` doesn't pan out, adding the vtable shim seems like an ad-hoc strategy that I'd like to be without.

Speaking of which, what is the progress of `unsized_locals`? In evaluating, it would be useful to know how close we are implementation wise to stabilizing?

> we can still pass self by-pointer into the method and clean up the memory for the Box after the fact

Can you elaborate on this? How and where would that be implemented? The nice thing about `unsized_locals` is that it is generally applicable and not ad-hoc; is this solution that as well?

**Comment by SimonSapin** [comment]
Checked. However the PR is still marked as "WIP". From the PR description:

> Major concern 1: instability

As already discussed in this thread, we don’t have a way to actually make `impl`s unstable. However https://github.com/rust-lang/rust/pull/55431#issuecomment-444228146 means agreement to land these as instantly stable, which resolves this concern.

Therefore, the `boxed_closure_impls` feature gate is fiction and should be removed from the PR.

`boxed-closure-impls.md` should also be removed, with perhaps its “Usage” section moved somewhere else. Maybe as a doc-comment on the impl?

I applaud the effort of writing up `fnbox.md`. However if landed it will likely be removed soon when `FnBox` is removed.

> Major concern 2: compatibility with FnBox

This is really not a major concern. `FnBox` is unstable, breakage of unstable features is to be expected. Keeping it functional as-is with a deprecation warning that says to use `Box<dyn FnOnce>` instead is sufficient, and not even mandatory. (Removing `FnBox` immediately would be acceptable, though less nice.)

The subtrait relationship and use of specialization are not necessary IMO. They involve removing the `FnBox::Output` associated type, which is another breaking change (though perhaps not one with much impact in practice). Still, since they’re already implemented I don’t really mind keeping them if others feel they are preferable.

> Minor concern: feature name and tracking issue

As mentioned above the instability is fiction, so there should be no tracking issue. The feature name (for `#[stable]` attributes) is fine.

File: src/doc/unstable-book/src/library-features/boxed-closure-impls.md [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+    let resource = "hello".to_owned();
+    // Create a boxed once-callable closure
+    let f: Box<dyn FnOnce(&i32)> = Box::new(|x| {
+        let s = resource;
+        println!("{}", x);
+        println!("{}", s);
+    });
+
+    // Call it
+    f();
```
Comment (crlf0710) [inline src/doc/unstable-book/src/library-features/boxed-closure-impls.md]: Doctest failure here. Maybe change this line to something like `f(&42);`.

**Comment by cramertj** [comment]
@Centril 
> > we can still pass self by-pointer into the method and clean up the memory for the Box after the fact

> Can you elaborate on this? How and where would that be implemented? The nice thing about unsized_locals is that it is generally applicable and not ad-hoc; is this solution that as well?

Sure! The key issue here is that the ABI for the function can't take `self` by-value directly. Instead, we need to generate a shim which takes `Box<Self>` and then calls the underlying function which takes `Self`. When we `CoerceUnsized` transform `Box<T: Sized + FnOnce(...) -> ...` into `Box<dyn FnOnce(...) -> ...>`, then we could have the compiler add vtable entries for `call_once` which take `self` by `Box<Self>`, dereference it, and call the underlying function. This obviously isn't generalizeable to all `self`-args-gated trait functions, so I'd expect it'd require a manual annotation on `call_once`. If you wish to generalize beyond `Box`, this technique could be similarly applied to all `DerefMove` types.

I'm not suggesting this as something that we should implement right now or that we should be doing instead, just noting that this is a possible route to implementing the functionality provided by this PR without relying on the `unsized_locals` feature.

**Comment by Centril** [comment]
@cramertj 

> [...]
> This obviously isn't generalizeable to all `self`-args-gated trait functions, so I'd expect it'd require a manual annotation on `call_once`. If you wish to generalize beyond `Box`, this technique could be similarly applied to all `DerefMove` types.

Ok, hmm... that attribute does seem like a rather ad-hoc way of extending the type system; we also don't have a notion of `DerefMove` as of yet so that's another thing the generalization would need to rely on.

> I'm not suggesting this as something that we should implement right now or that we should be doing instead, just noting that this is a possible route to implementing the functionality provided by this PR without relying on the `unsized_locals` feature.

Oh sure. :) My (temporary) concern amounts to that I wouldn't want to rely on an ad-hoc measure to back the impls in this PR if `unsized_locals` did not pan out.

**Comment by cramertj** [comment]
@Centril I found another way we could implement this that doesn't rely on `unsized_locals`: we could add a `call_box(self: Box<Self>)` function that delegates to the `call` function. I've got a working example in [this playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=9f1f8fe7e5ac10291a090551800a2a10). (pls someone tell me if this is something that has been tried before and doesn't work for some reason, as this seems like something we should've done years ago XD -- though it does rely on `specialization` to work)

**Comment by cramertj** [comment]
(this is also basically just a language-level way to spell the "vtable shim to take `&move`" thing I was talking about above)

**Comment by Centril** [comment]
Thanks @cramertj; You make a convincing case that the probability is good that we have sufficient not-super-ad-hoc backup measures should `unsized_locals` not pan out; therefore:

@rfcbot resolve risky-stabilization-order

**Comment by nikomatsakis** [comment]
@rfcbot reviewed

I'm late to the party. But I think I agree with @cramertj that we can, if forced, find another way to supply these impls. And I definitely agree that we ought to supply these impls eventually.

**Comment by alexreg** [comment]
This PR is still tagged "[WIP]", but with the FCP now begun, can I ask if it needs more work still, or can get merged right away upon completion of the FCP?

**Comment by qnighy** [comment]
I've just done a quick fix to resolve the conflict and removed the `[WIP]` label to avoid confusion. I'll work on fixes mentioned in https://github.com/rust-lang/rust/pull/55431#issuecomment-450024802 later (maybe after the decision to stabilize).

File: src/liballoc/boxed.rs [hunk 3]
```diff
@@ -738,6 +765,7 @@ impl<A, F> FnBox<A> for F
 #[unstable(feature = "fnbox",
            reason = "will be deprecated if and when `Box<FnOnce>` becomes usable", issue = "28796")]
 impl<A, R> FnOnce<A> for Box<dyn FnBox<A, Output = R> + '_> {
```
[the remark this answers is not in the document]
Comment (qnighy) [reply src/liballoc/boxed.rs]: I wanted to keep the `FnOnce<A> for Box<F>` impl gated behind `boxed_closure_impls` while the `FnOnce<A> for Box<dyn FnBox..>` impl is gated behind `fnbox`. It was wrong in two ways: impl stability is not checked, and we don't need to keep `fnbox` working as-is.

**Comment by qnighy** [comment]
> The subtrait relationship and use of specialization are not necessary IMO. They involve removing the `FnBox::Output` associated type, which is another breaking change (though perhaps not one with much impact in practice). Still, since they’re already implemented I don’t really mind keeping them if others feel they are preferable.

Given that we're going to insta-stabilize `impl FnOnce for Box<impl FnOnce + ?Sized>`, I'll remove the specialization. Still, the subtrait relation `FnBox: FnOnce` seems to be necessary to keep `FnBox` for a while. Otherwise we'll need to keep the `impl FnOnce for Box<dyn FnBox>` impl, but it conflicts with the former impl due to coherence involving generics.

**Comment by qnighy** [comment]
I've moved the old docs to https://gist.github.com/qnighy/fe17551217ec0415b7f21f928287e431.

**Comment by Centril** [comment]
Seems spurious... @bors retry

**Comment by bjorn3** [comment]
```
[01:05:03] LLVM ERROR: Cannot emit physreg copy instruction
```

**Comment by kennytm** [comment]
Sounds like #45417

**Comment by cramertj** [comment]
@kennytm Sorry, I see you marked this as waiting-on-author-- is there something that needs to be changed here, or is CI just flaking?

**Comment by kennytm** [comment]
@cramertj the error is https://github.com/rust-lang/rust/pull/55431#issuecomment-462742761 on the gnux32 target, and it doesn't look like an existing spurious error. Probably hit an LLVM bug. So this is waiting for the author (or anybody else) to debug why is this happening and how to fix/workaround it.

**Comment by qnighy** [comment]
Easily reproduced in my machine by setting `target = ["x86_64-unknown-linux-gnu", "x86_64-unknown-linux-gnux32"]`. No further progress yet.

**Comment by Dylan-DPC** [comment]
ping from triage @qnighy any updates?

**Comment by Dylan-DPC** [comment]
ping frm triage @qnighy 
Unfortunately we haven't heard from you on this in a while, so I'm closing the PR to keep things tidy. Don't worry though, if you'll have time again in the future please reopen this PR, we'll be happy to review it again! Thanks for contributing.

**Comment by cramertj** [comment]
Procedurally it seems a bit odd to me to close a perfectly fine PR due to it being unmergable as a result of happening to trigger a suspected LLVM bug. To be clear, there's no requested change to the code in this PR-- only a change to LLVM to successfully compile it.

**Comment by Centril** [comment]
@cramertj This is entirely [by the book](https://forge.rust-lang.org/triage-procedure.html):

> If the author’s been unresponsive for more than 14 days, close the PR due to inactivity and ask the author to reopen when they have a chance to make the necessary changes. Make sure to thank the author for the changes. Also tag the PR with S-inactive-closed.
>
> If there has been no meaningful updates after two triage updates, with no meaningful being defined as no commits or no status updates that show progress (i.e. “Soon TM”). The PR should be closed due to prolonged inactivity and ask the author to reopen when they have a chance to make the necessary changes. Make sure to thank the author for the changes, and warn them not to push to the PR while it is closed as GitHub will prevent the PR from being reopened. Make sure to also tag the PR with S-inactive-closed.

Once the LLVM issues can be fixed the PR should be reopened.

**Comment by cramertj** [comment]
> ask the author to reopen when they have a chance to make the necessary changes

This sounds like we're asking the author for changes. There are no changes being requested here.

**Comment by Centril** [comment]
Or generally actions (e.g. "fix LLVM") that unblocks the PR. Perhaps this should be labeled as `S-blocked-closed`.

**Comment by earthengine** [comment]
Are there any investigations have been done on LLVM? Can anyone find a workaround (maybe not optimal; we need it to be workable)?

**Comment by alexreg** [comment]
Has this bug been reported to LLVM? Does anyone have a link?

**Comment by kennytm** [comment]
@alexreg Please follow #59674 for the LLVM bug.

**Comment by alexreg** [comment]
So, it looks like this PR has effectively been implemented and merged in https://github.com/rust-lang/rust/pull/59500. Great!

**Pull request closed without merging**